### PR TITLE
Core: Port core functions to class

### PR DESCRIPTION
**Pull request by shreyans800755**
Port core functions to Session class.

Closes https://github.com/coala/coala/issues/4351

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: coalib/core/Core.py [hunk 1]
```diff
@@ -327,34 +147,172 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
+            :param bears:
+                The bear instances to run.
+            :param result_callback:
+                A callback function which is called when results are available.
+                Must have following signature::
+
+                    def result_callback(result):
+                        pass
+            """
+        cls.bears = bears
```
Comment (Makman2) [inline coalib/core/Core.py]: ahh no make it instance bound, that is way safer and then we can execute multiple sessions at once :D

Comment (Makman2) [inline coalib/core/Core.py]: So instantiate a new `Session` class passing those fields :)

File: tests/core/CoreTest.py [hunk 2]
```diff
@@ -317,7 +317,8 @@ def test_correct_reuse_of_implicitly_instantiated_dependency(self):
 
     def test_empty_case(self):
         # Test totally empty case.
-        dependency_tracker, bears_to_schedule = initialize_dependencies(set())
+        dependency_tracker, bears_to_schedule = initialize_dependencies(
+            set())
```
Comment (Makman2) [inline tests/core/CoreTest.py]: unrelated change ;)

File: tests/core/CoreTest.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 
 from coalib.settings.Section import Section
 from coalib.core.Bear import Bear
-from coalib.core.Core import initialize_dependencies, run
+from coalib.core.Core import Session, initialize_dependencies
```
Comment (Makman2) [inline tests/core/CoreTest.py]: if possible maintain alphabetical order of imports ;)
```
from coalib.core.Core import initialize_dependencies, Session
```

File: coalib/core/Core.py [hunk 4]
```diff
@@ -327,34 +147,172 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
+
+            :param bears:
+                The bear instances to run.
+            :param result_callback:
+                A callback function which is called when results are available.
+                Must have following signature::
+
+                    def result_callback(result):
+                        pass
+            """
```
Comment (Makman2) [inline coalib/core/Core.py]: misaligned docstring

File: coalib/core/Core.py [hunk 5]
```diff
@@ -327,34 +147,172 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
-    dependency_tracker, bears_to_schedule = initialize_dependencies(bears)
-
-    # Let's go.
-    schedule_bears(bears_to_schedule, result_callback, dependency_tracker,
-                   event_loop, {}, executor)
-    try:
-        event_loop.run_forever()
-    finally:
-        event_loop.close()
+class Session:
```
Comment (Makman2) [inline coalib/core/Core.py]: a class docstring would be good :+1:
also with a simple usage example :+1:

File: coalib/core/Core.py [hunk 6]
```diff
@@ -254,8 +76,6 @@ def initialize_dependencies(bears):
     dependencies are included so the dependency chain can be processed
     correctly.
 
-    :param bears:
-        The set of instantiated bears to run that serve as an entry-point.
```
Comment (Makman2) [inline coalib/core/Core.py]: why that change? the param is still there and needed :O

Comment (shreyans800755) [inline coalib/core/Core.py]: By mistake

File: coalib/core/Core.py [hunk 7]
```diff
@@ -327,34 +150,177 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
+        # FIXME   systems (for example Mesos).
+
+        # Set up event loop and executor.
+        self.event_loop = asyncio.SelectorEventLoop()
+        self.executor = concurrent.futures.ProcessPoolExecutor()
+        self.running_tasks = {}
+
+        # Initialize dependency tracking.
+        self.dependency_tracker, bears_to_schedule = initialize_dependencies(
+            self.bears)
```
Comment (Makman2) [inline coalib/core/Core.py]: maybe moving everything up to this statement into `__init__`? Traditionally member variables are set inside `__init__` to keep everything in one place.

File: coalib/core/Core.py [hunk 8]
```diff
@@ -327,34 +150,177 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
+        self.schedule_bears(bears_to_schedule)
+        try:
+            self.event_loop.run_forever()
+        finally:
+            self.event_loop.close()
+
+    def schedule_bears(self, bears):
+        """
+        Schedules the tasks of bears to the given executor and runs them on the
+        given event loop.
```
Comment (Makman2) [inline coalib/core/Core.py]: adapt docstrings ;) you don't have to pass any parameters except bears now.

-->
```
Schedules the tasks of bears.
```

File: coalib/core/Core.py [hunk 9]
```diff
@@ -327,34 +150,177 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
+                logging.debug('Scheduled {!r} (tasks: {})'.format(bear,
+                                                                  len(tasks)))
+
+                if not tasks:
+                    # We need to recheck our runtime if something is left to
+                    # process, as when no tasks were offloaded the event-loop
+                    # could hang up otherwise.
+                    self.cleanup_bear(bear)
+
+    def cleanup_bear(self, bear):
```
Comment (Makman2) [inline coalib/core/Core.py]: I'm thinking of prepending an underscore to the method names, to mark them as "private".
What do you think?

File: coalib/core/Core.py (hunk 7 above)
Comment (shreyans800755) [inline coalib/core/Core.py]: Good point. What about `executor' ? There comments in the code indicating that we'll allow custom executor in future. In that case, if we move `self.executor = concurrent.futures.ProcessPoolExecutor()` in constructor, then it will add additional cost if custom executor is used. I don't know how much expensive this call is.
If it is not much expensive, we can do it constructor. Otherwise, we can do have an param in `run` function called `executor` and code in the function should look something like `self.executor = executor if executor is not None else concurrent.futures.ProcessPoolExecutor()`.
But that change is not in scope of this PR.
What do you think ?

File: coalib/core/Core.py [hunk 10]
```diff
@@ -256,6 +78,7 @@ def initialize_dependencies(bears):
 
     :param bears:
         The set of instantiated bears to run that serve as an entry-point.
+
```
Comment (Makman2) [inline coalib/core/Core.py]: unnecessary change ;)

File: coalib/core/Core.py [hunk 11]
```diff
@@ -327,34 +150,177 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
 
 
-def run(bears, result_callback):
+class Session:
     """
-    Runs a coala session.
+    Maintains Session for a coala execution. For each session, there are set
+    of bears along with callback function, which is run when results are
+    available. On running a session, first its dependencies are initalized and
+    all the bears are scheduled to run in appropriate order.
```
Comment (Makman2) [inline coalib/core/Core.py]: a bit grammatical polishing ;)
-->
```
Maintains a session for a coala execution. For each session, there are set bears to run along with a callback function, which is run when results are available.
```
About this part:
```
On running a session, first its dependencies are initalized and
all the bears are scheduled to run in appropriate order.
```
I would remove that altogether, as 1. it's implementation detail 2. it's expected from the user that this works, so imo it requires no mention. And `all the bears are scheduled to run in appropriate order` is technically not completely correct ;) Bears that depend on another bear aren't scheduled until the bear dependency has finished. So in the beginning not all bears may be scheduled :)

File: tests/core/CoreTest.py [hunk 12]
```diff
@@ -383,7 +383,8 @@ def execute_run(bears):
         def on_result(result):
             results.append(result)
 
-        run(bears, on_result)
+        session = Session(bears, on_result)
+        session.run()
```
Comment (Makman2) [inline tests/core/CoreTest.py]: what do you think about providing a shortcut? So we place a `run` function into the core module, that actually does this:
```
def run(bears, on_result):
    session = Session(bears, on_result)
    session.run()
```
For users this means one statement less (avoiding the `Session` instantiation) and we can maybe add more functions later (if that makes sense of course).

Comment (Makman2) [inline tests/core/CoreTest.py]: If you want to do this (I think we should do this :D), then mention this inside the commit message :+1:

File: coalib/core/Core.py [hunk 13]
```diff
@@ -327,34 +150,177 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
 
-    :param bears:
-        The bear instances to run.
-    :param result_callback:
-        A callback function which is called when results are available. Must
-        have following signature::
+    Ex: If BearA requires BearB as dependencies, then, on running BearA,
+        first BearB will be executed, and after that, BearA will start running.
```
Comment (shreyans800755) [inline coalib/core/Core.py]: Its actually `Example`. I think you're right. I'll just remove it.

File: coalib/core/Core.py [hunk 14]
```diff
@@ -327,34 +150,177 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
+    def __init__(self, bears, result_callback):
+        """
+        :param bears:
+            The bear instances to run.
+        :param result_callback:
+            A callback function which is called when results are available.
+            Must have following signature::
 
             def result_callback(result):
                 pass
```
Comment (Makman2) [inline coalib/core/Core.py]: I think you need to indent this example too by 4 spaces :+1:

File: tests/core/CoreTest.py (hunk 12 above)
Comment (shreyans800755) [inline tests/core/CoreTest.py]: I would rather add an extra parameter in `execute_run()` function related to `on_result`. So, that it can be reused instead of having different functions.

File: coalib/core/Core.py (hunk 7 above)
Comment (Makman2) [inline coalib/core/Core.py]: Like you said, I think we will implement it by just using an additional parameter `executor`.
But separate PR / separate commit, as this is another issue :) (https://github.com/coala/coala/issues/4349)

File: coalib/core/Core.py [hunk 15]
```diff
@@ -327,34 +149,176 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
+
+    def run(self):
+        """
+        Runs a coala session.
+        """
+        # FIXME Allow to pass different executors nicely, for example to execute
+        # FIXME   coala with less cores, or to schedule jobs on distributed
+        # FIXME   systems (for example Mesos).
+
+        # Let's go.
```
Comment (Makman2) [inline coalib/core/Core.py]: I think this comment is not needed any more :P

File: coalib/core/Core.py [hunk 16]
```diff
@@ -327,34 +149,176 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
+        self.result_callback = result_callback
+
+        # Set up event loop and executor.
+        self.event_loop = asyncio.SelectorEventLoop()
+        self.executor = concurrent.futures.ProcessPoolExecutor()
+        self.running_tasks = {}
+
+        # Initialize dependency tracking.
+        self.dependency_tracker, self.bears_to_schedule =\
+            initialize_dependencies(self.bears)
```
Comment (Makman2) [inline coalib/core/Core.py]: please use braces instead of `\` :+1:

File: tests/core/CoreTest.py [hunk 17]
```diff
@@ -377,13 +377,17 @@ def setUp(self):
...
 
         def on_result(result):
             results.append(result)
 
-        run(bears, on_result)
+        if result_callback is None:
+            result_callback = on_result
+
+        session = Session(bears, result_callback)
+        session.run()
```
Comment (Makman2) [inline tests/core/CoreTest.py]: have you read my comment about providing a convenience function?

File: coalib/core/Core.py [hunk 18]
```diff
@@ -327,34 +149,176 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
+
+    def run(self):
+        """
+        Runs a coala session.
```
Comment (Makman2) [inline coalib/core/Core.py]: maybe `a` --> `the`?

File: coalib/core/Core.py [hunk 19]
```diff
@@ -327,34 +149,176 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
     """
-    Runs a coala session.
-
-    :param bears:
-        The bear instances to run.
-    :param result_callback:
-        A callback function which is called when results are available. Must
-        have following signature::
+    Maintains a session for a coala execution. For each session, there are set
+    of bears to run along with a callback function, which is run when results
```
Comment (Makman2) [inline coalib/core/Core.py]: ah, we should substitute this:

"which is ~~run~~ called when results are available."

Avoiding using `run` twice.

File: tests/core/CoreTest.py (hunk 12 above)
Comment (shreyans800755) [inline tests/core/CoreTest.py]: Oh I misunderstood your comment.

File: tests/core/CoreTest.py (hunk 17 above)
Comment (shreyans800755) [inline tests/core/CoreTest.py]: Misunderstood it on first go. Corrected. Additionally, I'm also returning `Session` object, just in case, if its needed further in code using `Session`.

**Comment by Asnelchristian** [comment]
As far as i am concern this is good, just address the few things i pointed out and we will merge this :tada: , marking WIP in the meantime :smile:

File: coalib/core/Core.py [hunk 20]
```diff
@@ -327,34 +149,192 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
+                    # FIXME   complete event-loop traceback.
+                    logging.error(
+                        'An exception was thrown during result-handling.',
+                        exc_info=ex)
+
+
 def run(bears, result_callback):
     """
-    Runs a coala session.
+    It initiates a session with the given parameters and runs it.
```
Comment (Makman2) [inline coalib/core/Core.py]: ~~It initiates~~ Initiates a session with ...

:smile:

File: tests/core/CoreTest.py [hunk 21]
```diff
@@ -514,7 +517,7 @@ def test_run_result_handler_exception(self):
         on_result = unittest.mock.Mock(side_effect=ValueError)
 
         with self.assertLogs(logging.getLogger()) as cm:
-            run({bear}, on_result)
+            self.execute_run({bear}, on_result)
```
Comment (Makman2) [inline tests/core/CoreTest.py]: is that change necessary?

Comment (Makman2) [inline tests/core/CoreTest.py]: (also the above ones)

File: coalib/core/Core.py [hunk 22]
```diff
@@ -327,34 +149,192 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
     return dependency_tracker, bears
 
 
+class Session:
+    """
+    Maintains a session for a coala execution. For each session, there are set
+    of bears to run along with a callback function, which is called when
+    results are available.
+
+    If BearA requires BearB as dependencies, then, on running BearA,
```
Comment (Makman2) [inline coalib/core/Core.py]: as **dependency**

Comment (Makman2) [inline coalib/core/Core.py]: you maybe want to prepend some general statement before the example, like
```
Dependencies of bears (provided via ``bear.BEAR_DEPS``) are automatically handled.
If BearA requires ...
```

Comment (Makman2) [inline coalib/core/Core.py]: you maybe want to prepend some general statement before the example, like

Dependencies of bears (provided via ``bear.BEAR_DEPS``) are automatically handled.
If BearA requires ...

File: coalib/core/Core.py [hunk 23]
```diff
@@ -327,34 +149,192 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
...
+
+
 def run(bears, result_callback):
     """
-    Runs a coala session.
+    It initiates a session with the given parameters and runs it.
 
     :param bears:
-        The bear instances to run.
+            The bear instances to run.
```
Comment (Makman2) [inline coalib/core/Core.py]: it's even unnecessary, the additional indentation is not needed^^ (and also inconsistent)

File: tests/core/CoreTest.py (hunk 21 above)
Comment (shreyans800755) [inline tests/core/CoreTest.py]: Not really. This is just to make consistent usage of execute_run every time.

Comment (shreyans800755) [inline tests/core/CoreTest.py]: So, that every time, we just have to call `execute_run` even if the definition of run change in future.

Comment (Makman2) [inline tests/core/CoreTest.py]: actually then this is a good test we should keep :)

Comment (Makman2) [inline tests/core/CoreTest.py]: but if you want to supply a custom callback, really use `run` directly. `execute_run` was written to be able to retrieve the results for assertion. Putting a custom callback into `execute_run` turns this into a simple `run(bear, custom_callback)` call (which is done here)

Comment (Makman2) [inline tests/core/CoreTest.py]: However if you still want it, please file an issue :)

Comment (shreyans800755) [inline tests/core/CoreTest.py]: Nah, I think its better to remove it. Doesn't make sense to put an unnecessary overhead.

**Comment by Makman2** [comment]
https://github.com/coala/coala/pull/4525 takes over.
